compiler/rustc_codegen_llvm/src/intrinsic.rs

```diff
@@ -1067,8 +1064,34 @@ fn generic_simd_intrinsic(
         let i1xn = bx.trunc(i_xn_msb, bx.type_vector(bx.type_i1(), in_len));
         // Bitcast <i1 x N> to iN:
         let i_ = bx.bitcast(i1xn, bx.type_ix(in_len));
-        // Zero-extend iN to the bitmask type:
-        return Ok(bx.zext(i_, bx.type_ix(expected_int_bits)));
+
+        match ret_ty.kind() {
+            ty::Uint(i) if i.bit_width() == Some(expected_int_bits) => {
+                // Zero-extend iN to the bitmask type:
+                return Ok(bx.zext(i_, bx.type_ix(expected_int_bits)));
+            }
+            ty::Array(elem, len)
+                if matches!(elem.kind(), ty::Uint(ty::UintTy::U8))
+                    && len.try_eval_usize(bx.tcx, ty::ParamEnv::reveal_all())
+                        == Some(expected_bytes) =>
+            {
+                // Zero-extend iN to the array lengh:
+                let ze = bx.zext(i_, bx.type_ix(expected_bytes * 8));
+
+                // Convert the integer to a byte array
+                let ptr = bx.alloca(bx.type_ix(expected_bytes * 8), Align::ONE);
+                bx.store(ze, ptr, Align::ONE);
```

Comment:
For anyone more familiar with this than me: is there any reason this alignment or the one below must be greater than a byte?

Reply:
Well, when these are handled by SSE or AVX512 instructions, they are normal integers of 16 to 64 bits. Even though we are saying we want to **interpret** this as a byte array, we should probably use the natural word-alignment of the processor: that way, `kmovq mem, reg` (reg to mem) naturally hits the alignment to pick it up in a GPR correctly, and if we for some reason are executing AVX512 features on a 32-bit processor, we still are correctly aligned for such (since then it would go into two GPRs).

Reply:
I don't think anything is preventing LLVM from using a greater alignment if it's more efficient, but the actual alignment is just 1, right?  Regular `u8` arrays are definitely used with larger alignments in most situations, but their alignment is still 1.

Reply:
As long as you're coherent about this (I believe `PlaceRef` tracks it for you, so it may be worth trying to use it more), you can increase the alignment of the `alloca` (and then pass that alignment to the `store`/`load` - whether manually, or relying on the `PlaceRef` API, or letting LLVM infer it).

I'm not sure how useful it is here, but it may make LLVM more likely to generate the aligned versions of instructions that have both aligned and unaligned variants (assuming LLVM doesn't rewrite this cast away).

Also, if you're using `alloca`s, please consider using `place.storage_{live,dead}()` before/after all of the memory operations on the `alloca` - again I'm unsure whether this is significant in the context (since e.g. a small wrapper that gets inlined will do the right thing itself during LLVM inlining), but the default without those (i.e. without `llvm.lifetime.{start,end}`) is every `alloca` adds to the stack frame size (for the entire function).

Reply:
Hmmm. At least hypothetically this should in fact be "a small wrapper that gets inlined". I don't think either of us is familiar with the PlaceRef API tho'. But that shouldn't be a blocking concern.